pkg/util/codec/codec.go

```diff
@@ -423,6 +424,175 @@ const (
 	KeepVarColumnLength
 )
 
+// PreAllocForSerializedKeyBuffer estimates key length
+func PreAllocForSerializedKeyBuffer(buildKeyIndex []int, chk *chunk.Chunk, tps []*types.FieldType, usedRows []int, filterVector []bool, nullVector []bool, serializeModes []SerializeMode, serializedKeysVectorBuffer [][]byte, memoryUsagePerRow []int64) ([]int64, error) {
+	rowNum := len(usedRows)
+	if cap(memoryUsagePerRow) < rowNum {
+		memoryUsagePerRow = make([]int64, rowNum)
+	} else {
+		memoryUsagePerRow = memoryUsagePerRow[:rowNum]
+		for i := range rowNum {
+			memoryUsagePerRow[i] = 0
+		}
+	}
+
+	for i, idx := range buildKeyIndex {
+		column := chk.Column(idx)
+		canSkip := func(index int) bool {
+			if column.IsNull(index) {
+				nullVector[index] = true
+			}
+			return (filterVector != nil && !filterVector[index]) || (nullVector != nil && nullVector[index])
+		}
+
+		switch tps[i].GetType() {
+		case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear:
+			for j := range memoryUsagePerRow {
+				memoryUsagePerRow[j] += size.SizeOfByte + 8
```

Comment:
Only if `SerializeMode` is `NeedSignFlag`, we need extra byte otherwise, 8 is enough